pkg/gpu/consumer.go

```diff
@@ -0,0 +1,166 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package gpu
+
+import (
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/gpu/model"
+	ddebpf "github.com/DataDog/datadog-agent/pkg/ebpf"
+	gpuebpf "github.com/DataDog/datadog-agent/pkg/gpu/ebpf"
+	"github.com/DataDog/datadog-agent/pkg/process/monitor"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// CudaEventConsumer is responsible for consuming CUDA events from the eBPF probe, and delivering them
+// to the appropriate stream handler.
+type CudaEventConsumer struct {
+	eventHandler   ddebpf.EventHandler
+	once           sync.Once
+	closed         chan struct{}
+	streamHandlers map[model.StreamKey]*StreamHandler
+	wg             sync.WaitGroup
+}
+
+// NewCudaEventConsumer creates a new CUDA event consumer.
+func NewCudaEventConsumer(eventHandler ddebpf.EventHandler) *CudaEventConsumer {
+	return &CudaEventConsumer{
+		eventHandler:   eventHandler,
+		closed:         make(chan struct{}),
+		streamHandlers: make(map[model.StreamKey]*StreamHandler),
+	}
+}
+
+// Stop stops the CUDA event consumer.
+func (c *CudaEventConsumer) Stop() {
+	if c == nil {
+		return
+	}
+	c.eventHandler.Stop()
+	c.wg.Wait()
+	c.once.Do(func() {
+		close(c.closed)
+	})
+}
+
+// Start starts the CUDA event consumer.
+func (c *CudaEventConsumer) Start() {
+	if c == nil {
+		return
+	}
+	health := health.RegisterLiveness("gpu-tracer-cuda-events")
+	processMonitor := monitor.GetProcessMonitor()
+	cleanupExit := processMonitor.SubscribeExit(c.handleProcessExit)
+
+	c.wg.Add(1)
+	go func() {
+		processSync := time.NewTicker(5 * time.Second)
+
+		defer func() {
+			cleanupExit()
+			err := health.Deregister()
+			if err != nil {
+				log.Warnf("error de-registering health check: %s", err)
+			}
+			c.wg.Done()
+			log.Debugf("CUDA event consumer stopped")
+		}()
+
+		dataChannel := c.eventHandler.DataChannel()
+		lostChannel := c.eventHandler.LostChannel()
+		for {
+			select {
+			case <-c.closed:
+				return
+			case <-health.C:
+			case <-processSync.C:
+				c.checkClosedProcesses()
+			case batchData, ok := <-dataChannel:
+				if !ok {
+					return
+				}
+
+				if len(batchData.Data) < gpuebpf.SizeofCudaEventHeader {
+					log.Errorf("Not enough data to parse header, data size=%d, expecting at least %d", len(batchData.Data), gpuebpf.SizeofCudaEventHeader)
```

Comment:
is that an error or a warn? (are we expecting that to happen often?)

Reply:
It's an error, it shouldn't happen at all.